sklearn/naive_bayes.py

```diff
@@ -1528,3 +1541,600 @@ def _joint_log_likelihood(self, X):
+    """
+    with _print_elapsed_time(message_clsname, message):
+        return estimator.partial_fit(X, y, **fit_params)
+
+
+def _jll_one(estimator, X):
+    """Call ``estimator.predict_joint_log_proba``.
+
+    See :func:`sklearn.pipeline._transform_one`.
+    """
+    return estimator.predict_joint_log_proba(X)
+
+
+class ColumnwiseNB(_BaseNB, _BaseComposition):
+    """
+    Column-wise Naive Bayes meta-estimator.
+
+    This estimator combines various naive Bayes estimators by applying them
+    to different column subsets of the input and joining their predictions
+    according to the naive Bayes assumption. This is useful when features are
+    heterogeneous and follow different kinds of distributions.
+
+    Parameters
+    ----------
+    nb_estimators : list of tuples
+        List of (name, nb_estimator, columns) tuples specifying the naive Bayes
+        estimators to be combined into a single naive Bayes meta-estimator.
+
+        name : str
+            Name of the naive Bayes estimator. Like in
+            :class:`~sklearn.pipeline.Pipeline`,
+            :class:`~sklearn.pipeline.FeatureUnion`,
+            and :class:`~sklearn.compose.ColumnTransformer`, this allows the
+            subestimator and its parameters to be set using :term:`set_params`
+            and searched in grid search.
+        nb_estimator : estimator
+            The estimator must support :term:`fit` or :term:`partial_fit`,
+            depending on how the meta-estimator is fitted. In addition, the
+            estimator must support ``predict_joint_log_proba`` method, which
+            takes :term:`X` of shape (n_samples, n_features) and returns a
+            numpy array of shape (n_samples, n_classes) containing joint
+            log-probabilities, ``log P(x,y)`` for each sample point and class.
+        columns :  str, array-like of str, int, array-like of int, \
+                array-like of bool, slice or callable
+            Indexes the data on its second axis. Integers are interpreted as
+            positional columns, while strings can reference DataFrame columns
+            by name.  A scalar string or int should be used where
+            ``nb_estimator`` expects X to be a 1d array-like (vector),
+            otherwise a 2d array will be passed to the transformer.
+            A callable is passed the input data `X` and can return any of the
+            above. To select multiple columns by name or dtype, you can use
+            :obj:`~sklearn.compose.make_column_selector`.
+
+    priors : array-like of shape (n_classes,) or str, default=None
+        Prior probabilities of classes. If unspecified, the priors are
+        calculated as relative frequencies of classes in the training data.
+        If str, the priors are taken from the estimator with the given name.
+        If array-like, the same priors might have to be specified manually in
+        each sub-estimator, in order to ensure consistent predictions.
```

Comment:
It does not feel right to task this particular meta-estimator with assigning hyperparameters (priors) to sub-estimators. My understanding of the workflow is that the user initialises sub-estimators with desired hyperparameters and passes the instances to the meta-estimator; here priors are just another hyperparameter. It is expected that meta-estimators do not mess with hyperparameters (even in clones), unless it is their purpose, as with GridSearchCV. Does this perspective make sense to you?

I cannot think of an example or reason for using different priors in sub-estimators either, and the few books that I had at hand did not discuss such a possibility at all. But the issue is a bit deeper than just assigning hyperparameters. As I mentioned in item 13 of the head post, I combine subestimators at `Log P(x,y)` using the formula that assumes equal priors and uses the meta-estimator's prior as the value for it. In detail, we expand the familiar naive Bayes postulate

` Log P(x|y) = Log P(x_1|y) + ... + Log P(x_N|y)`

to obtain 

`Log P(x,y) - Log P(y) = Log P(x_1,y) - Log P(y) + ... + Log P(x_N,y) - Log P(y)`
				
where `Log P(y)` on the right-hand side come from the respective sub-estimators. Assuming all `Log P(y)` are equal, we get

`Log P(x,y) = Log P(x_1,y) + ... + Log P(x_N,y) - (N - 1) * Log P(y)`

which is the currently used formula. You might wonder, why do we use this formula instead of the one we started with? First, it was proposed and seconded in superseded issues and pull requests. Second, `Log P(x,y)` were available in _BaseNB and derived classes. Third, `Log P(x|y)` is at risk of conditioning on a zero prior; this risk increases if we collect conditional probabilities or priors from each sub-estimator, as the initial formula proposes. To sum up, equal priors are assumed in the model, but are not checked or required in the API. Any thoughts? After thinking about this for some time, I concluded that I am overthink this.